scripts/prepdocslib/textsplitter.py

```diff
@@ -16,16 +18,92 @@ def split_pages(self, pages: List[Page]) -> Generator[SplitPage, None, None]:
             yield  # pragma: no cover - this is necessary for mypy to type check
 
 
+STANDARD_LINE_BREAKS = [",", ";", ":", " ", "(", ")", "[", "]", "{", "}", "\t", "\n"]
+CJK_LINE_BREAKS = [
+    "。",
+    "、",
+    "，",
+    "；",
+    "：",
+    "（",
+    "）",
+    "【",
+    "】",
+    "「",
+    "」",
+    "『",
+    "』",
+    "〔",
+    "〕",
+    "〈",
+    "〉",
+    "《",
+    "》",
+    "〖",
+    "〗",
+    "〘",
+    "〙",
+    "〚",
+    "〛",
+    "〝",
+    "〞",
+    "〟",
+    "〰",
+    "〾",
+    "〿",
+    "–",
+    "—",
+    "‘",
+    "’",
+    "‚",
+    "‛",
+    "“",
+    "”",
+    "„",
+    "‟",
+    "‹",
+    "›",
+    "‽",
+    "⁇",
+    "⁈",
+    "⁉",
+    "⸮",
+    "ⸯ",
+    "。",
+    "々",
+    "〻",
+    "〼",
+]
+cl100k = tiktoken.get_encoding("cl100k_base")
+
+
+def split_page_by_max_tokens(page_num: int, text: str, max_tokens: int) -> Generator[SplitPage, None, None]:
+    """
+    Splits page by tokens
+    """
+    tokens = cl100k.encode(text)
+    if len(tokens) < max_tokens:
+        yield SplitPage(page_num=page_num, text=text)
+    else:
+        # Split page in half and call function again
+        # TODO : Overlap first and second half
+        first_half = text[: len(text) // 2]
```

Comment:
FWIW we should never split + encode cl100k tokens because 1 token doesn't always produce a Unicode character, in CJK and other languages you need 2 or 3 tokens for a single character.